### ISSUE TITLE
DebugMode reports incorrect redistribute placement in backwards

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#167655]
### 🐛 Describe the bug

Source: https://colab.research.google.com/drive/1zugXJm47oX4LOGfczk1H0xKs9d_Cn-sm#scrollTo=ahfnmPO3gtS3

Repro:

```
import torch
from torch import Tensor
from torch.distributed.tensor import (
    DTensor,
    DeviceMesh,
    distribute_tensor,
    init_device_mesh,
    Partial,
    Replicate,
    Shard,
)
from torch.distributed.tensor.placement_types import _StridedShard
from torch.distributed._local_tensor import (
    local_tensor_mode,
    LocalTensor,
    LocalTensorMode,
)
import traceback

S = Shard
R = Replicate()
_SS = _StridedShard

def product(it):
    x = 1
    for i in it:
        x *= i
    return x

def arange_nd(*sizes):
    if len(sizes) == 1 and isinstance(sizes[0], (list, tuple)):
        sizes = sizes[0]
    return torch.arange(product(sizes)).view(sizes)

def reconcile(l: Tensor):
    """Asserts that a LocalTensor is the same on all ranks, and returns the single Tensor."""
    if isinstance(l, LocalTensor):
        return l.reconcile()
    return l

def init_local_tensor_mode(world_size):
    from torch.distributed import _local_tensor
    if _local_tensor._LOCAL_TENSOR_MODE:
        for lm in list(reversed(_local_tensor._LOCAL_TENSOR_MODE)):
            lm.__exit__(None, None, None)
    try:
        torch.distributed.destroy_process_group()
    except AssertionError:
        pass
    torch.distributed.init_process_group(
        "fake",
        rank=0,
        world_size=world_size,
    )
    lm = LocalTensorMode(world_size)
    lm.__enter__()
    return world_size

# https://github.com/jax-ml/jax/pull/29381

world_size = init_local_tensor_mode(4)
mesh = init_device_mesh("cpu", (2, 2), mesh_dim_names=("x", "y"))

a = distribute_tensor(arange_nd(8, 4).float(), mesh, [S(0), S(1)]).requires_grad_()
b = distribute_tensor(arange_nd(4, 6).float(), mesh, [R, S(0)]).requires_grad_()
c = distribute_tensor(arange_nd(8, 4).float(), mesh, [S(0), S(1)]).requires_grad_()
d = distribute_tensor(arange_nd(4, 6).float(), mesh, [R, S(0)]).requires_grad_()
e = a @ b
f = c @ d
g = e + f
h = g.redistribute(placements=[S(0), R])  # comms!
i = h.sum()

e.retain_grad()
f.retain_grad()
g.retain_grad()
h.retain_grad()
i.retain_grad()

from torch.utils._debug_mode import DebugMode
with DebugMode(record_realtensor=False) as debug_mode:
    i.backward()
print(debug_mode.debug_string())

"""
print(a.grad.placements)
print(b.grad.placements)
print(c.grad.placements)
print(d.grad.placements)
print(e.grad.placements)
print(f.grad.placements)
print(g.grad.placements)
print(h.grad.placements)
print(i.grad.placements)
"""
```

This prints:

```
  aten::ones_like(dt: f32[]| PR, pin_memory=False, memory_format=torch.preserve_format)
  aten::clone(dt: f32[]| RR, memory_format=torch.contiguous_format)
  aten::expand(dt: f32[]| RR, [8, 6])
  aten::clone(dt: f32[8, 6]| RR, memory_format=torch.contiguous_format)
    redistribute_input(t: f32[8, 6], trace: RR->S(0)R->S(0)P)
  aten::clone(dt: f32[8, 6]| S(0)R, memory_format=torch.contiguous_format)
  aten::clone(dt: f32[8, 6]| S(0)R, memory_format=torch.contiguous_format)
  aten::t(dt: f32[8, 4]| S(0)S(1))
  aten::mm(dt: f32[4, 8]| S(1)S(0), dt: f32[8, 6]| S(0)R)
  aten::t(dt: f32[4, 6]| RS(0))
  aten::mm(dt: f32[8, 6]| S(0)R, dt: f32[6, 4]| RS(1))
  aten::detach(dt: f32[8, 4]| S(0)S(1))
  aten::detach(dt: f32[4, 6]| PS(0))
  aten::clone(dt: f32[8, 6]| S(0)R, memory_format=torch.contiguous_format)
  aten::t(dt: f32[8, 4]| S(0)S(1))
  aten::mm(dt: f32[4, 8]| S(1)S(0), dt: f32[8, 6]| S(0)R)
  aten::t(dt: f32[4, 6]| RS(0))
  aten::mm(dt: f32[8, 6]| S(0)R, dt: f32[6, 4]| RS(1))
  aten::detach(dt: f32[8, 4]| S(0)S(1))
  aten::detach(dt: f32[4, 6]| PS(0))
```

In particular, look at the redistribute:

```
    redistribute_input(t: f32[8, 6], trace: RR->S(0)R->S(0)P)
```

It claims that the tensor is redistributed to partial. But it is not, in DTensor, we have hard-coded that you should stay replicate when this happens.

```
        output = redistribute_local_tensor(
            local_tensor,
            current_spec,
            previous_spec,
            async_op=async_op,
            is_backward=True,
        )

        if output.dtype != ctx.original_dtype:
            output = output.to(ctx.original_dtype)

        # normalize the target placement to replicate if it is partial
        normalized_placements: list[Placement] = []
        for previous_placement in previous_spec.placements:
            if previous_placement.is_partial():
                # keep target placement to replicate instead of partial in this case
                normalized_placements.append(Replicate())
            else:
                normalized_placements.append(previous_placement)

```

But because the redistribute hook is in redistribute_local_tensor it reports the P placement (prior to it getting normalized).

The "right" fix is to stop hacking this up with the normalization after the fact, and ask for the redistribute we actually want.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @zpcore 

### Versions

main

**Comment by zpcore** [comment]
Let me draft a PR to approach the issue
